### PR TITLE
Generate trait help instead of hardcoding

### PR DESCRIPTION
**Pull request by bumbadadabum**
Introduces a new key "help_text=" in [trait]

### CODE REVIEW

**Comment by Vultraz** [comment]
This is a very good change.

**Comment by bumbadadabum** [comment]
I'll add some extra things to this tomorrow.

**Comment by Wedge009** [comment]
I concur with vultraz - I think there's a bit of duplication with regards to the help text.

**Comment by gfgtdf** [comment]
What exactly happend when a trit has no help_text= attribute?

**Comment by CelticMinstrel** [comment]
Availability is an enumeration, so it absolutely should not be translated. Showing a different translated string based on its value could be an option though.

**Comment by bumbadadabum** [comment]
@gfgtdf I actually didn't think of that, good point. Also, when a trait has no help_text it defaults to the trait's description. For the availability, I could perhaps make it "Must-have" or "Unavailable" if it's musthave or none.

**Comment by bumbadadabum** [comment]
Ok I did that. Tell me if there are any more complaints.

**Comment by gfgtdf** [comment]
> Also, when a trait has no help_text it defaults to the trait's description. 

Note that the description as shown in the sidebar is automatically generated in most cases. It be nice if the code would also fallback to that in those cases (from the code it looks like you'd just use the discription= attribute in [trait]). This is mostlikeley a nontrivial task since the effects implementation now goes through lua and the game_lua_kernel is not available on titlescreen (in case people watch the help on the tiltescreen). We can also think about that later, so it's not really a blocker.

File: src/help/help_impl.cpp [hunk 1]
```diff
@@ -623,6 +625,65 @@ std::vector<topic> generate_faction_topics(const config & era, const bool sort_g
...
+	for (const config & trait : unit_types.traits()) {
+		const std::string trait_id = trait["id"];
+		trait_list.insert(std::make_pair(trait_id, trait));
+	}
+
+
+	for (const unit_type_data::unit_type_map::value_type &i : unit_types.types())
+	{
+		const unit_type &type = i.second;
+		if (description_type(type) == FULL_DESCRIPTION) {
```
Comment (CelticMinstrel) [inline src/help/help_impl.cpp]: Is this the check for whether the unit has been seen or the pref is set to show all units?

**Comment by CelticMinstrel** [comment]
I wonder if it's really a good idea to show the trait's availability though... it could have a different value depending on race or unit type. It's good to show this in the race or unit type description (if traits are listed there), but it doesn't feel like it belongs in the general trait description.

About the description, I don't think it should use the generated portion (especially since these go through the Lua kernel(, but from the code it looks like it already isn't, so that should be fine, I think.

**Comment by bumbadadabum** [comment]
availability= is supplied in the [trait] tag, so I figured it'd be safe to put in here as well.

The check I added for checking if the unit is already discovered is the same one as the one used in the generation for ability topics, so I figured this would take the setting into account. 

@gfgtdf The main reason I decided against using the generated portion is mostly the fact that the generated portion is integrated with the unit itself, and would require an entire rewrite (plus look worse than in the unit sidebar). I think using a help_text key would be the easiest way to allow UMC creators to also give their traits some nice descriptions, and going through the effort of rewriting the trait effect text generation for just a fallback thing only potentially used in UMC doesn't seem worth it to me.

**Pull request merged**